cmd/snap-confine/mount-support.h

```diff
@@ -41,4 +41,16 @@ void sc_populate_mount_ns(const char *base_snap_name, const char *snap_name);
  * ensure that "/snap" is mounted shared. See LP:#1668659
  */
 void sc_ensure_shared_snap_mount();
+
+/**
+ * Set up user mounts, private to this process.
+ *
+ * If any user mounts have been configured for this process, this does
+ * the following:
+ * - create a new mount namespace
+ * - reconfigure all existing mounts to slave mode
+ * - perform all user mounts
+ */
+void sc_setup_user_mounts(const char *snap_name);
```

Comment:
How do you intend to marry this with `snap-update-ns`? Ideally snap-update-ns would be completely responsible for performing profile-based mount operations. I think this is desirable because it would allow us to keep one implementation of the logic (the one in golang), not two.

In my mind the ideal flow would be:
- snap-confine sets up basic stuff that is hard to to from go and does unshare and pivot_root
- snap-update-ns is invoked in system mode, it processes the system wide mount profiles
- snap-confine captures the system mode namespace and unshares again 
- snap-update-ns is invoked in user mode, it processes the per-user mount profiles
- snap-confine captures the user mode namespace and invokes snap-exec

WDYT?

Reply:
So with my current implementation, we don't capture the second name space we create: it is effectively private to this particular invocation of the command, and will go away when it exits.  This was intended as a way to ensure that these mounts wouldn't persist past the end of the user's session.

This means that any costs associated with setting up the mounts will be incurred every time a command is run rather than being amortised over multiple runs.  Does this make any difference over whether snap-update-ns would be appropriate?

One other thing that might affect the decision: I was discussing some of this with Gustavo yesterday, and he mentioned that he wanted to have a system where namespaces would be automatically discarded when the last instance of a command from a snap exited (although it wasn't quite clear how this would be achieved).  If something like that existed, then maybe persisting per-user namespaces would make sense, since we'd know that they'd go away at the end of the session.

Reply:
> One other thing that might affect the decision: I was discussing some of this with Gustavo yesterday, and he mentioned that he wanted to have a system where namespaces would be automatically discarded when the last instance of a command from a snap exited (although it wasn't quite clear how this would be achieved). If something like that existed, then maybe persisting per-user namespaces would make sense, since we'd know that they'd go away at the end of the session.

I have a rough implementation of that that uses the freezer cgroup to detect empty vs non-empty snap process sets. This is not something that can actively detect the group becoming empty but can do so on demand.